Examples/CustomMachine/jax_wrapper.py

```diff
@@ -0,0 +1,168 @@
+        so that it can be used as a NetKet machine.
+
+        Args:
+            hilbert: Hilbert space on which the state is defined. Should be a
+                subclass of `netket.hilbert.Hilbert`.
+            module: A pair `(init_fn, predict_fn)`. See the documentation of
+                `jax.experimental.stax` for more info.
+            seed: Seed to use to construct `jax.random.PRNGKey` for
+                initialisation of parameters. If `None` seed will be chosen
+                automatically (which is __not__ synchronized between MPI
+                processes so prefer `self.init_random_parameters` if you're
+                using multiple MPI tasks)
+        """
+        # NOTE: The following call to __init__ is important!
+        super(JAXMachine, self).__init__(hilbert)
+        if seed is None:
+            # NOTE(twesterhout): I believe, jax uses 32-bit integers internally
+            # to represent the seed. Hence the limit
+            seed = random.randint(0, 2 ** 32 - 1)
+        init_fn, self._forward_fn = module
+        self._forward_fn = jax.jit(self._forward_fn)
+        input_shape = (-1, hilbert.size)
+        output_shape, self._params = init_fn(jax.random.PRNGKey(seed), input_shape)
+        if output_shape != (-1, 2):
+            raise ValueError("module's output_shape is weird, check your network")
+        # Computes total number of parameters
+        n_par = sum(reduce(lambda n, p: n + p.size, layer, 0) for layer in self._params)
+        # Save the initial dtype (for debugging mostly)
+        dtypes = set(p.dtype for layer in self._params for p in layer)
+        if len(dtypes) != 1:
+            raise ValueError("module parameters have different dtypes")
+        self._dtype = next(iter(dtypes))
+        self._complex_dtype = {
+            np.dtype("float32"): np.complex64,
+            np.dtype("float64"): np.complex128,
+        }[self._dtype]
+        assert all(
+            np.asarray(p).flags.c_contiguous for layer in self._params for p in layer
+        ), "sorry, column major order is not supported (yet)"
+        self._n_par = lambda: n_par
+        # Computes the Jacobian matrix using backprop
+        self._jacobian = jax.jacrev(self._forward_fn)
+        self._jacobian = jax.jit(self._jacobian)
+
+    def _log_val(self, x, out):
+        """
+        Do not use this function directly! It's an implementation detail!
+
+        See `self.log_val`.
+        """
+        out[:] = (
+            np.asarray(self._forward_fn(self._params, x))
+            .view(dtype=self._complex_dtype)
+            .squeeze()
+        )
+
+    def _der_log(self, x, out):
+        """
+        Do not use this function directly! It's an implementation detail!
+
+        See `self.log_val`.
+        """
+        J = self._jacobian(self._params, x)
+        batch_size = x.shape[0]
+        i = 0
+        for g in (g.reshape(batch_size, 2, -1) for layer in J for g in layer):
```

Comment:
I don't understand what `this for in for in for in` loop does, maybe it's me... ahha